clients/geth/specular/rollup/utils/hook.go

```diff
@@ -0,0 +1,132 @@
+
+type RollupConfig interface {
+	GetCoinbase() common.Address
+	GetL2ChainID() uint64
+	GetL1FeeOverhead() uint64
+	GetL1FeeMultiplier() float64
+	GetL1OracleAddress() common.Address
+}
+
+func SpecularEVMHook(msg types.Message, evm *vm.EVM, cfg RollupConfig) error {
+	tx := createTransaction(msg, cfg)
+	fee, err := calculateL1Fee(tx, evm, cfg)
+	if err != nil {
+		return err
+	}
+	return chargeL1Fee(fee, msg, evm, cfg)
+}
+
+func createTransaction(msg types.Message, cfg RollupConfig) *types.Transaction {
+	var txData types.TxData
+	if msg.GasTipCap() != nil {
+		txData = &types.DynamicFeeTx{
+			ChainID:    big.NewInt(int64(cfg.GetL2ChainID())),
+			Nonce:      msg.Nonce(),
+			GasTipCap:  msg.GasTipCap(),
+			GasFeeCap:  msg.GasFeeCap(),
+			Gas:        msg.Gas(),
+			To:         msg.To(),
+			Value:      msg.Value(),
+			Data:       msg.Data(),
+			AccessList: msg.AccessList(),
+		}
+	} else if msg.AccessList() != nil {
+		txData = &types.AccessListTx{
+			ChainID:    big.NewInt(int64(cfg.GetL2ChainID())),
+			Nonce:      msg.Nonce(),
+			GasPrice:   msg.GasPrice(),
+			Gas:        msg.Gas(),
+			To:         msg.To(),
+			Value:      msg.Value(),
+			Data:       msg.Data(),
+			AccessList: msg.AccessList(),
+		}
+	} else {
+		txData = &types.LegacyTx{
+			Nonce:    msg.Nonce(),
+			GasPrice: msg.GasPrice(),
+			Gas:      msg.Gas(),
+			To:       msg.To(),
+			Value:    msg.Value(),
+			Data:     msg.Data(),
+		}
+	}
+	return types.NewTx(txData)
+}
+
+func calculateL1Fee(tx *types.Transaction, evm *vm.EVM, cfg RollupConfig) (uint64, error) {
+	// calculate L1 gas from RLP encoding
+	buf := new(bytes.Buffer)
+	err := tx.EncodeRLP(buf)
+	if err != nil {
+		return 0, err
+	}
+	bytes := buf.Bytes()
+	// remove the last 3 bytes containing the signature
+	// this mirrors the optimism implementation [1]
+	// but contradicts the optimism spec [2]
+	// [1] https://github.com/ethereum-optimism/optimism/blob/5d9a38dcd9dc79dce41a6d08f9b28ff850f77811/l2geth/rollup/fees/rollup_fee.go#L204
+	// [2] https://github.com/ethereum-optimism/optimism/blob/develop/specs/exec-engine.md#l1-cost-fees-l1-fee-vault
+	rlp := bytes[:len(bytes)-3]
+
+	zeroes, ones := zeroesAndOnes(rlp)
+	txDataGas := zeroes*TX_DATA_ZERO + ones*TX_DATA_ONE + cfg.GetL1FeeOverhead()
+	basefee := readBasefee(evm, cfg)
+
+	l1Fee := txDataGas * basefee
+	scaledL1Fee := uint64(math.Ceil(float64(l1Fee) * cfg.GetL1FeeMultiplier()))
+
+	return scaledL1Fee, nil
+}
+
+func chargeL1Fee(L1Fee uint64, msg types.Message, evm *vm.EVM, cfg RollupConfig) error {
+	senderBalance := evm.StateDB.GetBalance(msg.From())
+	if senderBalance.Uint64() < L1Fee {
+		return errors.New("insufficient balance to cover L1 fee")
+	}
+
+	evm.StateDB.AddBalance(cfg.GetCoinbase(), big.NewInt(int64(L1Fee)))
+	evm.StateDB.SubBalance(msg.From(), big.NewInt(int64(L1Fee)))
+
+	log.Info("charged L1 Fee", "fee", L1Fee)
+	return nil
+}
+
+func readBasefee(evm *vm.EVM, cfg RollupConfig) uint64 {
```

Comment:
I'd make a generalized function for reading L1 oracle slots (i.e. pass the slot hash in the argument)